tests/realtikvtest/txntest/stale_read_test.go

```diff
@@ -1506,3 +1493,245 @@ func TestStaleReadNoBackoff(t *testing.T) {
+	safePointComment := "All versions after safe point can be accessed. (DO NOT EDIT)"
+	updateSafePoint := fmt.Sprintf(`INSERT INTO mysql.tidb VALUES ('%[1]s', '%[2]s', '%[3]s')
+	ON DUPLICATE KEY
+	UPDATE variable_value = '%[2]s', comment = '%[3]s'`, safePointName, safePointValue, safePointComment)
+	tk.MustExec(updateSafePoint)
+
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (id int primary key, v int)")
+
+	// Insert row #1
+	tk.MustExec("insert into t values (1, 10)")
+	time.Sleep(1000 * time.Millisecond)
+	firstTime := time.Now().Add(-500 * time.Millisecond)
+	// Retrieve current TSO from store's Oracle instead of @@tidb_current_ts.
+	externalTS, err := store.GetOracle().GetTimestamp(context.Background(), &oracle.Option{})
+	if err != nil {
+		t.Fatalf("failed to get TSO: %v", err)
+	}
+
+	time.Sleep(1000 * time.Millisecond)
+	// Insert row #2
+	tk.MustExec("insert into t values (2, 20)")
+	time.Sleep(1000 * time.Millisecond)
+	secondTime := time.Now().Add(-500 * time.Millisecond)
+
+	staleReadMethods := []struct {
+		name   string
+		setup  func()
+		query  string
+		clean  func()
+		expect []string
+	}{
+		{
+			name: "tidb_read_staleness",
+			setup: func() {
+				tk.MustExec("set @@tidb_read_staleness='-2'")
+			},
+			query: "select * from t",
+			clean: func() {
+				tk.MustExec("set @@tidb_read_staleness=''")
+			},
+			expect: []string{"1 10"},
+		},
+		{
+			name:   "AS OF TIMESTAMP sees only first row",
+			setup:  func() {},
+			query:  fmt.Sprintf("SELECT * FROM t AS OF TIMESTAMP '%s'", firstTime.Format("2006-1-2 15:04:05.000")),
+			clean:  func() {},
+			expect: []string{"1 10"},
+		},
+		{
+			name:   "AS OF TIMESTAMP sees both rows (using secondTime)",
+			setup:  func() {},
+			query:  fmt.Sprintf("SELECT * FROM t AS OF TIMESTAMP '%s'", secondTime.Format("2006-1-2 15:04:05.000")),
+			clean:  func() {},
+			expect: []string{"1 10", "2 20"},
+		},
+		/*
+			We cannot test TIDB_BOUNDED_STALENESS here because it relies
+			on the update of safe ts, which cannot be controlled by the test
+			{
+				name:   "AS OF TIMESTAMP with TIDB_BOUNDED_STALENESS",
+				setup:  func() {},
+				query:  "select * from t as of timestamp tidb_bounded_staleness(TIMESTAMPADD(SECOND, -5, NOW()), TIMESTAMPADD(SECOND, -1, NOW()))",
+				clean:  func() {},
+				expect: []string{"1 10", "2 20"},
+			},
+		*/
+		{
+			name: "SET TRANSACTION",
+			setup: func() {
+				tk.MustExec(fmt.Sprintf("SET TRANSACTION READ ONLY AS OF TIMESTAMP '%s'", firstTime.Format("2006-1-2 15:04:05.000")))
+			},
+			query:  "select * from t",
+			clean:  func() {},
+			expect: []string{"1 10"},
+		},
+		{
+			name: "external ts read",
+			setup: func() {
+				tk.MustExec("set @@tidb_enable_external_ts_read=1")
+				// ignore the failure when setting external_ts, as it reports error if not set to a larger value
+				_, _ = tk.Exec(fmt.Sprintf("set @@global.tidb_external_ts=%d", externalTS))
```

Comment:
I didn't understand it. Does it mean that it returns error while it's actually successful to se it?

Reply:
In the combination test the setting is executed multiple times. When setting to a same value(namely not a larger one) the SQL returns error, which we do not care about.

Reply:
I'm afraid it would make us unable to identify the problem when there's some other problem causing the statement failed. Can we consider using a increasing timestamp in this case, as it seems we have a 500ms gap of timestamps free to use in this test, so that the error can be avoided?

Reply:
That makes sense. Modified.